index.js

```diff
@@ -13,89 +13,107 @@
-        credentials = require(prefix+"/runtime/nodes/credentials");
+        this._address = '127.0.0.1';
+        this._listenPort = 0; // ephemeral
 
-    } catch (err) {
-        // ignore, assume init will be called again by a test script supplying the runtime path
+        this.init();
     }
-}
-
-initRuntime(runtimePath);
-
-var app = express();
-
-var address = '127.0.0.1';
-var listenPort = 0; // use ephemeral port
-var port;
-var url;
-var logSpy;
-var server;
 
-function helperNode(n) {
-    RED.nodes.createNode(this, n);
-}
+    _initRuntime(requirePath) {
+        try {
+            const RED = this._RED = require(requirePath);
+
+            // public runtime API
+            this._redNodes = RED.nodes;
+            this._events = RED.events;
+            this._log = RED.log;
+
+            // access internal Node-RED runtime methods
+            const prefix = path.dirname(requirePath);
+            this._context = require(path.join(prefix, 'runtime', 'nodes', 'context'));
+            this._comms = require(path.join(prefix, 'api', 'editor', 'comms'));
+
+            this.credentials = require(path.join(prefix, 'runtime', 'nodes', 'credentials'));
+
+            // proxy the methods on Node.prototype to both be Sinon spies and asynchronously emit
+            // information about the latest call
+            const NodePrototype = require(path.join(prefix, 'runtime', 'nodes', 'Node')).prototype;
+            PROXY_METHODS.forEach(methodName => {
+                const spy = this._sandbox.spy(NodePrototype, methodName);
+                NodePrototype[methodName] = new Proxy(spy, {
+                    apply: (target, thisArg, args) => {
+                        const retval = Reflect.apply(target, thisArg, args);
+                        process.nextTick(function(call) { return () => {
+                            NodePrototype.emit.call(thisArg, `call:${methodName}`, call);
+                        }}(spy.lastCall));
+                        return retval;
+                    }
+                });
+            });
+        } catch (ignored) {
+            // ignore, assume init will be called again by a test script supplying the runtime path
+        }
+    }
 
-module.exports = {
-    init: initRuntime,
-    load: function(testNode, testFlow, testCredentials, cb) {
-        var i;
+    init(runtimePath = findRuntimePath()) {
+        if (runtimePath) {
+            this._initRuntime(runtimePath);
+        }
+    }
 
-        logSpy = sinon.spy(log,"log");
+    load(testNode, testFlow, testCredentials, cb) {
+        const log = this._log;
+        const logSpy = this._logSpy = this._sandbox.spy(log, 'log');
         logSpy.FATAL = log.FATAL;
         logSpy.ERROR = log.ERROR;
         logSpy.WARN = log.WARN;
         logSpy.INFO = log.INFO;
         logSpy.DEBUG = log.DEBUG;
         logSpy.TRACE = log.TRACE;
         logSpy.METRIC = log.METRIC;
+        this.log = () => logSpy;
```

Comment:
suggest that instead of assigning `this.log = () => logSpy;` here, that we make the object interface more explicit by moving it outside the code. i.e. use `function log() { return this._logSpy; }` in the class.